{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "# RNA-Seq Differential Expression (DE) Analysis Using baySeq\n",
    "\n",
    "Thomas J. Hardcastle (2019). baySeq: Empirical Bayesian analysis of\n",
    "  patterns of differential expression in count data. R package version\n",
    "  2.18.0.\n",
    "The original paper from Hardcastle & Kelly introducing the concepts implemented in this package is available here \n",
    "https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-422\n",
    "\n",
    "We will be using the pasilla package for our example data which is cited here.\n",
    " Wolfgang Huber and Alejandro Reyes (2019). pasilla: Data package with\n",
    "  per-exon and per-gene read counts of RNA-seq samples of Pasilla\n",
    "  knock-down by Brooks et al., Genome Research 2011.. R package version\n",
    "  1.12.0.\n",
    "  \n",
    "Data for the package is part of this paper: \n",
    "\"Conservation of an RNA regulatory map between Drosophila and mammals\" by Brooks AN, Yang L, Duff MO, Hansen KD, Park JW, Dudoit S, Brenner SE, Graveley BR, Genome Res. 2011 Feb;21(2):193-202, Epub 2010 Oct 4, PMID: 20921232. \n",
    "\n",
    "Briefly, baySeq uses empirical Bayesian inference to determine the likelihood that genes in compared samples are indeed DE genes. baySeq improve seeks to improve accuracy in DE estimation over other popular packages by using the underlying structure of the data itself. baySeq does show improved performance in the case of more complex study designs (i.e. multiple group comparisons) and in studies with large numbers of libraries compared to other popular packages. Here we illustrate a simplified example of a comparison between a control and a treatment group.\n",
    "\n",
    "Imagine you want to place a bet on whether it will rain later today. What kind of information do you think would help you make the most accurate bet, a coin toss or accounting for humidity, cloud cover, or season? Bayesian inference represents the latter."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Definition of terms\n",
    "\n",
    "###### Bayesian statistics: \n",
    "Learning from experience to make inferences about the relationship between our variables of interest. While the null hypothesis in the frequentist approach (i.e. the use of p values) says that \"there is no relationship\", the Bayesian approach allows us to incorporate previous knowledge we may have about a relationship into future research we want to do on it. The associated credible interval (as compared to the confidence interval) tells us how likely it is that a particular value of interest lies between the prior and posterior estimates of that value. In our example we would interpret this as, given what we know about genetic expression between our control and treatment groups, how likely it is that a certain percentage of genes are DE genes?\n",
    "<br>\n",
    "###### Prior probability: \n",
    "The information we have about our data before we see it. This could come from what we know about similar experiments or we can estimate it from our data and these options vary in their utility. In our example, this could refer to how likely we think it is that a certain percentage of genes are DE between our control and treatment groups.\n",
    "<br>\n",
    "###### Posterior probability:\n",
    "The results of incorproating our prior knowledge and our observed data. In our case that means, given that our control and treatment samples have particular genes with particular amounts of expression, how likely it is that a certain percentage of genes between them are DE genes?\n",
    "\n",
    "For more information and the source for these definitions please see \"A Gentle Introduction to Bayesian Analysis\" from Schoot, Kaplan, Denissen, Asendorpf, Neyer, and Aken (Developmental Method, 2013) https://onlinelibrary.wiley.com/doi/full/10.1111/cdev.12169"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our very first step is to load the baySeq and pasilla packages from Bioconductor use an R cluster to parallelize our work. Please see http://bioconductor.org/ for information about installation and use of Bioconductor and its packages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(baySeq)\n",
    "library (pasilla)\n",
    "if(require(\"parallel\")) cl <- makeCluster(8) else cl <- NULL"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load example dataset\n",
    "Our example data are from the pasilla package available on Bioconductor. The experiment studied RNAi knockdown of Pasilla, the Drosophila melanogaster ortholog of mammalian NOVA1 and NOVA2, on the transcriptome. Data are provided by NCBI Gene Expression Omnibus under accession numbers GSM461176 to GSM461181.\n",
    "\n",
    "Please reach out to EICC if you would like to compare 3 or more groups as this is a simplified example. It may also be the case you will need more than 6 samples per experimental group or that you may need to remove genes with average counts greater than 5, 10, 15, or even 20 for sufficient statistical power. Please see our PROPER walkthrough for an example of our of power and sample size analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "countdata <- system.file(\"extdata\",\n",
    "                      \"pasilla_gene_counts.tsv\",\n",
    "                      package=\"pasilla\", mustWork=TRUE)\n",
    "\n",
    "countdata <- as.matrix(read.csv(countdata,sep=\"\\t\",row.names=\"gene_id\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Labelling replicates and specifying our hypothesis\n",
    "<br>\n",
    "After that, we label our samples correctly and specify our hypothesis. \n",
    "\n",
    "In our case, our replicates are organized in our csv file above such that the first 4 columns are untreated samples and the latter 3 columns are treated samples. This is how we label the replicates in the model.\n",
    "\n",
    "The groups object specifies our hypothesis. In the null hypothesis of no differential expression (NDE), if all of our samples are 1's, this means that all of the samples belonging to the same group, there is no difference between the treated and untreated samples. In the alternative hypothesis of differential expression, (DE) 1's represent membership in a single group and 2's represent membership in a second group. This set of 4 1's and 3 2's says that there exist two distinct patterns of expression in the control and treatment groups."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "replicates <- c(\"Untreated\",\"Untreated\",\"Untreated\",\"Untreated\",\n",
    "                \"Treated\",\"Treated\",\"Treated\")\n",
    "\n",
    "groups<- list(NDE = c(1,1,1,1,1,1,1),\n",
    "               DE = c(1,1,1,1,2,2,2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preparation for Differential Expression Analysis\n",
    "In order to perform our pairwise comparison we need to specify some informaton about our data. Replicates and Groups represent the labels for our replicates (Control vs. Treatment) and the null (NDE) and alternative (DE) hypotheses we specified above. We also annotate all 13064 genes with a number to cross-reference with the list of gene names and counts after analysis. Our final preparation step is to get our library size. Sometimes you may already know this but we have inferred it from the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "CD <- new(\"countData\", data = countdata, replicates = replicates, groups = groups)\n",
    "libsizes(CD) <- getLibsizes(CD)\n",
    "CD@annotation <-data.frame(name = rownames(countdata))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determine prior probabilities\n",
    "<br>\n",
    "We determine the prior probabilities of DE from the count data. Recall that this means we are incorporating some structure inherent in our data to inform the likeliehood that there are DE genes between our two groups. \n",
    "<br>\n",
    "<br>\n",
    "For our simplified example we have chosen to do 10000 bootstrap samples using quasi-likelihood estimation.These parameters will vary by experiment and we would tailor these to your specific project.The length of time that determining the priors and posterior likelihoods will take depends on your computational power, your sample size, the method of estimation, the number of genes, number of samples, and the number of groups you are comparing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Finding priors...done.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Time difference of 4.277674 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "start_time <- Sys.time()\n",
    "CD <- getPriors.NB(CD, samplesize = 10000, estimation = \"QL\", cl = cl)\n",
    "end_time <- Sys.time()\n",
    "\n",
    "end_time - start_time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determine posterior likelihoods\n",
    "<br>\n",
    "Next, we determine the posterior likelihood, that is, we are using our prior knowledge and looking at our existing data to estimate how likely is it that there are DE genes between our two groups?\n",
    "\n",
    "We chose to use just 3 bootstrap estimates here in the interest of the accuracy vs. time trade-off. We would assess this parameter based on your specific experiment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Length of priorReps:10010\n",
      "Length of priorSubset:14599\n",
      "Length of subset:14599\n",
      "Length of postRows:14599\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "..."
     ]
    },
    {
     "data": {
      "text/plain": [
       "Time difference of 13.29088 mins"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "start_time <- Sys.time()\n",
    "CD <- getLikelihoods(CD, bootStraps = 3, verbose = FALSE, cl = cl)\n",
    "end_time <- Sys.time()\n",
    "end_time - start_time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summarize Results\n",
    "\n",
    "Here we have the list of top 10 genes which are most likely to be differentially expressed at FDR controlled at 0.05. This cut-off may vary anywhere from 0.01-0.2 and will depend on the individual experiment but should be specified a priori. We can see that the very first gene, FBgn002511 is significantly up regulated in the treated as compared to the untreated samples whereas FBgn0039155 is significantly up regulated in the untreated samples as compared to the treated samples.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<caption>A data.frame: 6 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>name</th><th scope=col>untreated1</th><th scope=col>untreated2</th><th scope=col>untreated3</th><th scope=col>untreated4</th><th scope=col>treated1</th><th scope=col>treated2</th><th scope=col>treated3</th><th scope=col>likes</th><th scope=col>DE</th><th scope=col>FDR.DE</th><th scope=col>FWER.DE</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>2366</th><td>FBgn0025111</td><td> 451</td><td> 653</td><td>279</td><td>306</td><td>4186</td><td>2381</td><td>2686</td><td>1.0000000</td><td>2&gt;1</td><td>2.479510e-09</td><td>2.479510e-09</td></tr>\n",
       "\t<tr><th scope=row>9831</th><td>FBgn0039155</td><td>1527</td><td>2220</td><td>741</td><td>931</td><td><span style=white-space:pre-wrap>  70</span></td><td><span style=white-space:pre-wrap>  38</span></td><td><span style=white-space:pre-wrap>  52</span></td><td>1.0000000</td><td>1&gt;2</td><td>5.009191e-09</td><td>1.001838e-08</td></tr>\n",
       "\t<tr><th scope=row>6948</th><td>FBgn0035085</td><td>1176</td><td>1579</td><td>651</td><td>785</td><td> 245</td><td> 135</td><td> 150</td><td>1.0000000</td><td>1&gt;2</td><td>9.268608e-09</td><td>2.780582e-08</td></tr>\n",
       "\t<tr><th scope=row>10305</th><td>FBgn0039827</td><td> 478</td><td> 678</td><td>319</td><td>353</td><td><span style=white-space:pre-wrap>  35</span></td><td><span style=white-space:pre-wrap>  14</span></td><td><span style=white-space:pre-wrap>  29</span></td><td>1.0000000</td><td>1&gt;2</td><td>1.448238e-08</td><td>5.792952e-08</td></tr>\n",
       "\t<tr><th scope=row>6711</th><td>FBgn0034736</td><td> 366</td><td> 645</td><td>286</td><td>271</td><td><span style=white-space:pre-wrap>  47</span></td><td><span style=white-space:pre-wrap>  27</span></td><td><span style=white-space:pre-wrap>  29</span></td><td>0.9999996</td><td>1&gt;2</td><td>8.748318e-08</td><td>4.374159e-07</td></tr>\n",
       "\t<tr><th scope=row>11422</th><td>FBgn0051092</td><td><span style=white-space:pre-wrap>  64</span></td><td> 103</td><td> 38</td><td> 39</td><td> 436</td><td> 220</td><td> 243</td><td>0.9999995</td><td>2&gt;1</td><td>1.619840e-07</td><td>9.719035e-07</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 12\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & name & untreated1 & untreated2 & untreated3 & untreated4 & treated1 & treated2 & treated3 & likes & DE & FDR.DE & FWER.DE\\\\\n",
       "  & <fct> & <int> & <int> & <int> & <int> & <int> & <int> & <int> & <dbl> & <fct> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t2366 & FBgn0025111 &  451 &  653 & 279 & 306 & 4186 & 2381 & 2686 & 1.0000000 & 2>1 & 2.479510e-09 & 2.479510e-09\\\\\n",
       "\t9831 & FBgn0039155 & 1527 & 2220 & 741 & 931 &   70 &   38 &   52 & 1.0000000 & 1>2 & 5.009191e-09 & 1.001838e-08\\\\\n",
       "\t6948 & FBgn0035085 & 1176 & 1579 & 651 & 785 &  245 &  135 &  150 & 1.0000000 & 1>2 & 9.268608e-09 & 2.780582e-08\\\\\n",
       "\t10305 & FBgn0039827 &  478 &  678 & 319 & 353 &   35 &   14 &   29 & 1.0000000 & 1>2 & 1.448238e-08 & 5.792952e-08\\\\\n",
       "\t6711 & FBgn0034736 &  366 &  645 & 286 & 271 &   47 &   27 &   29 & 0.9999996 & 1>2 & 8.748318e-08 & 4.374159e-07\\\\\n",
       "\t11422 & FBgn0051092 &   64 &  103 &  38 &  39 &  436 &  220 &  243 & 0.9999995 & 2>1 & 1.619840e-07 & 9.719035e-07\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 12\n",
       "\n",
       "| <!--/--> | name &lt;fct&gt; | untreated1 &lt;int&gt; | untreated2 &lt;int&gt; | untreated3 &lt;int&gt; | untreated4 &lt;int&gt; | treated1 &lt;int&gt; | treated2 &lt;int&gt; | treated3 &lt;int&gt; | likes &lt;dbl&gt; | DE &lt;fct&gt; | FDR.DE &lt;dbl&gt; | FWER.DE &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 2366 | FBgn0025111 |  451 |  653 | 279 | 306 | 4186 | 2381 | 2686 | 1.0000000 | 2&gt;1 | 2.479510e-09 | 2.479510e-09 |\n",
       "| 9831 | FBgn0039155 | 1527 | 2220 | 741 | 931 |   70 |   38 |   52 | 1.0000000 | 1&gt;2 | 5.009191e-09 | 1.001838e-08 |\n",
       "| 6948 | FBgn0035085 | 1176 | 1579 | 651 | 785 |  245 |  135 |  150 | 1.0000000 | 1&gt;2 | 9.268608e-09 | 2.780582e-08 |\n",
       "| 10305 | FBgn0039827 |  478 |  678 | 319 | 353 |   35 |   14 |   29 | 1.0000000 | 1&gt;2 | 1.448238e-08 | 5.792952e-08 |\n",
       "| 6711 | FBgn0034736 |  366 |  645 | 286 | 271 |   47 |   27 |   29 | 0.9999996 | 1&gt;2 | 8.748318e-08 | 4.374159e-07 |\n",
       "| 11422 | FBgn0051092 |   64 |  103 |  38 |  39 |  436 |  220 |  243 | 0.9999995 | 2&gt;1 | 1.619840e-07 | 9.719035e-07 |\n",
       "\n"
      ],
      "text/plain": [
       "      name        untreated1 untreated2 untreated3 untreated4 treated1 treated2\n",
       "2366  FBgn0025111  451        653       279        306        4186     2381    \n",
       "9831  FBgn0039155 1527       2220       741        931          70       38    \n",
       "6948  FBgn0035085 1176       1579       651        785         245      135    \n",
       "10305 FBgn0039827  478        678       319        353          35       14    \n",
       "6711  FBgn0034736  366        645       286        271          47       27    \n",
       "11422 FBgn0051092   64        103        38         39         436      220    \n",
       "      treated3 likes     DE  FDR.DE       FWER.DE     \n",
       "2366  2686     1.0000000 2>1 2.479510e-09 2.479510e-09\n",
       "9831    52     1.0000000 1>2 5.009191e-09 1.001838e-08\n",
       "6948   150     1.0000000 1>2 9.268608e-09 2.780582e-08\n",
       "10305   29     1.0000000 1>2 1.448238e-08 5.792952e-08\n",
       "6711    29     0.9999996 1>2 8.748318e-08 4.374159e-07\n",
       "11422  243     0.9999995 2>1 1.619840e-07 9.719035e-07"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(topCounts(CD, group=\"DE\", normaliseData=TRUE, FDR=0.05))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualization\n",
    "<br>\n",
    "Finally, we can graph what we see. Here is one example of a plot we could make for you called an MA-plot where the x axis is the mean gene expression and the y axis is the log2 transformed fold change. Truly DE genes are in red and non-DE genes are in black."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlgAAAJYCAMAAACJuGjuAAAANlBMVEUAAABNTU1oaGh8fHyM\njIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6enw8PD/AAD/pQD///+afi2oAAAACXBIWXMA\nABJ0AAASdAHeZh94AAAgAElEQVR4nO2di3rauhJGlTZNuts0Onr/lz3B0lx0832MIP/6WiBg\nS7a1GI1kAy4AYIC79waA5wRiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIB\nEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjAB\nYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIBEyAW\nMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgET\nIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFi\nARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYw\nAWIBEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMg\nFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIB\nEyAWMAFiARMgFjABYgETIBYwAWIBE+4glnOu9ejr8fS3U3jXYXrBe59W6FcxU7vmq6gT9gwI\nl4s1tWt1E61y6uEiN6tcVCvksjSemn9psgpqncr1YvGtfsS3ErR8EbRU9Erx6mZgFCvfDReq\np+Zf8uoWnMPVYnHflz0ir6hHvD3wdO+VTulxNCqaFUqzXHHfrF3ji3twAhALYpkAsSCWCcix\nAnIsCzAqDBgVWoB5rAnMY50NZt6BCRALmACxgAl7xXp7P3UzwLOxV6yYNZ+6KeCZ2C/WP4gF\n+ux141c+/D91m8ATsFeJz1eIBWY4ogR8Al0gFjABbgATIBYw4ZBY7z+QvIM2R5R4x6gQ9Dii\nxIv7fdp2gCfDaFSo57j+978Q8P+x/++QY7dWIby6T+sqwKNypNX/vfz8a1wFeFSOdYWrkneI\n9R2BWMCEC1odYn1HIBYwAWIBE/ZfQbr6eiyI9R2BWMAEdIXABIgFTIBYwASIBUyAWMAEiAVM\ngFjABIgFTIBYwASIBUyAWMAEiAVMgFjABIgFTIBYwASIBUyAWMAEiAVMgFjABIgFTIBYwASI\nBUyAWMAEiAVMgFjAhIOt/vuHc69/TKu4AHw77/kc+Vm5EH7GD9i/mVRxFTO/TQ52c0isN/f2\nGcK/t/lvTx69yZy6BWdxSKyX+O22n+6HRRUX4Yp7cAqHxKL+o+5HHug35yCWCYfE+kVivVhU\ncREQy4T9Yr2+//7j/vt6+Pk2n72P3mLIsSw4+sVr08OX2R8SGL3JMCq0YPfx/Pj4/fv1dUrh\n3+Z/oGL8JjsvDxw+o7wMzLyfCGKfALFOBNmaALFunNODYXypgFjn9WAQSwGxzuvBIJYCYp3o\nA3IsAWKdKRZGhQzEOrUHwzwW8ShiWbbY5h4M+izzGGLZ9jEbS0eHt4YHEeusgnrlb/EEKfoa\nHkKsocbxQ23MuECsrQy1MeMCsbYy1MaMy0OINVZaM9TGDMuDiDXSQGyojRmWxxBrsKmjoTZm\nUB5FrLsBifbxAGLds2nR7e1leLHu27RI1PcyvlgnlHGwcpi1g9HFum/T9mpH4rUIxNpeOxKv\nFTylWCd+TrBVORKvFQwvVnRkSxlnBpRWWUi81jC2WF9NOn2OP2vapXh0bkCpa4NYaxhdrKhW\n9lSYiUfee/N2h1hrGFoskigTa6ZM7+nmULWLm2VZ/rMMOMcWK9268qlOoZNSzl4su1Hh8ww4\nRxar9YWBc2KRUf5QrWswCyvPM+AcWqzGcXYzw0Qt1mO+658ofRtaLFf1DNMzRdbFeLp/2DwF\nYl1QRfPLcV0yq5O8q9uFojdsxdpFTwBi2VfhWnNYKYL1GjsbFc6UHNZ2lZfn0sixzKtIqVTe\nqo1hYo7va8VFtRuvaevl7YxRoXUVrpFgzXyv/Iry+Ka1We0GvUfP9LD5YcG4Yn3Fnt7JnB0l\nyoptsZrlPlHKczmDinUb2Lnb6RklFp/e2ROw5L5lS8cgiLWfQcWi8WAxN+qWxoVL2yAhqxGw\neiELXu1gTLF8Y67BiVM0YNy+DZJppfFl/XK22tpcur01z5Iv7WFMsdid9lRDmdfnY8HmL0M5\nKTdbP8vaWuPCdVqFhn/PM8Lbw5BitedG060rT/Xks1fNS7jyMFUUGKpXNm+v3py0A9+9Hx1U\nLF/IIddlcWfIIuTz7U79KwpV0UlJy1MYuw9F2ZG2/P12jCjWNCLUAYtSKw5lOijxGUKqzOn7\n1obkPenxI6CcVyXeX6y7pngDihXduV0LmqfcPE6M/QzptUcsuV23efNNVFyP6NSzK8u34M4p\n3rBiFZ2V9oojWLag59Q+rdMXi+pZuXW9JuK+NRO1E7guZ4jqR6rCtcQqu0H92It2qddse5Ul\n6qHvS2+1aju5unZStTtknNKF3bsnHlksVUCeYWUi3cyiE0CBFWy2aD55tXYurJPds25UqKtf\n2adV2OdjUUxxfzXDiVUHLJqYkn9B5rimP73zahIiX7kqPYROCJrdfu6Cy71ydROePHdxpBiI\nJUtzcNJPSM+ng1egBg90EySAzNYR1rV8mbM1xeKIWqy3j9OEQI6VL1wGLO758u5Q/12KtcKb\nVS2vRnva2nx91Ttv2c9+pcX9/oLO6VJ3Vz9YFVVH6EIZs8rH1B3R7eJ7dY0DkjSpkSitKW3G\neV57tLC93jO7sLNc31f5WFWk+SsdsNQJ6cqtLFo0lCzxlOMvHPJyBlXqDqxQYdfijq6NIPnb\n4q5yHGEwsSarvPqYzWQC2aRmsxqBS+479caTik7l77yJrREfz1I5iUwhy9ZpiVX5zNqcRwt4\n5+7sCGOJlcKTFquMTyxRHqF0t9gVa1rOB4k6acGy+ZQvKUpJtFJLurBBLHZ/xVEogi7EOlwF\nHX0/5xVHrzwhu/09dXXU/mXh8QPSt3NApGa2gZVYQadvKuuqxGpUWAnEPfaW+HNivnU5Q4nF\n4syKlVvGzeVit9nPoqaANc16cS6XbZ+sIL7oISYnV8rEdo/V2ACWFGLdoQpO1H35RB6t6MF0\ngR/Fq5sxblql1ePEZaeWjZdvOb6Gq9F83MVxPyt3TovDsTKrsO7B9Eobjke9ZQ/DSGI5HgKK\nWL6TZE2v3WTy3OrxkYvfkFW09GTg1E8Gti8En3VhrayL75XOISuet6fe4dxUVWbtfe+IlOU8\nDgOJFWOQ0x0hjxFbZsWuzFOylT7ZE0/wlH2Tj8Ulu1wKXkGHLF6aZKZnyasgz+abndfVDYH8\njmj31NmB6Jb9MAwmVsx96Ik0RqSJrdwsusg08y566QtVnFeFp8W0WLr5uDjeKJV5t/am2e/J\nPYW4QMkZ6dU/DGprMI91vIo8HN3gjCs9qZeI/sirotjkFtWcGjEvieczpMvVMYLMU9sVpKmL\nnrCZBzkeQkr/R24uploP3P1pRhIr5L1NpYPEpSkd86oHTPGIFvYpVafU2qdQ5Z0Wx3lduyTs\nIaSBQGAZaHYjz5bi7EbgFeNjT1leyCyhqimAZYdF7fW2ozYsI4nlZHw/PcEdWGoT7516JvZn\nvJBLAnH7kXjBxYT+67/nZCllb1J3oKxe5qxYyxDUn2y+qz4dNN2kRI7eJvFF+XxaOamaVx8g\n1ulVlCcJZYzoKTAV01qc2oesl2zMm0aHvJweCo6mHyQDCoHE4tEf3eaJN0emeEuRLRXhp3I8\n1+yquQ2nltdHCGJZVKEGgLRaUk0FIWVLUqvK6332vLLDq4UDrxX0Ui5On8pK7J0qia4o5LGr\n6vecC2q8ySv5fATKe1AeIOmM3cN7NY5Y1NlJ5u6dskGLkw8PlVCkGzvKL5IVDYK+D7KiJN9B\nR7bUNYqB6QxRkOFA1IjDHuVhZBa5HNeovpk+HxXq2h+KQcRSeTqtNTlV5+0tqXKTuAPtJv+Z\nY0VAUksEFbLkVUrwadI/9cXaHhfSCck4csi6zjyfanwzfRWxGqZ1DvRI/g0nFj0Rz/s1pWjL\nVi7ZdjJPx1S/mGumxFL9pPwvRAtihbqEIhXgUxafRaz42Hne+36O5Yq/2wfZrfbvGsYQS4KL\n/ubjgmZH1jepLiHoEV+j+NTvStemsqFcq7Q8PwqSOcmJJk70PXeSKipxGHTaw1ws5dv8YZQx\nBcTKF+GhX3qCOrY1Aaug2QH6NDkRK8te4ap0h8qJVwiUU2V5flYKTas7GoF65YxzNC8hi6qA\nF9J+q9oc9a2qL144jE4NN9vH93LhhhBreo97mgIIFMF6Z6BXUcUvGnE2lvLFQ93qQUJLrgu5\nkBmhl1G9ZZpfozLj+U3pLNMaKfCELFStEcsV/+vDe30nOYpY6dwvPUFz6G2z9GR7I7SVubyv\nph/KkngJ6t/yDrNSy5fbFkK+gH6pUaAjsbIAKcGRbQ1sxELAmhVrZTOcyhhilYmrxI/ZmOVV\nqMmf943l5ouh09OZgmruQS/v9LMh6ws1of6LvJGz4vplFa/qfnvpCKvusH38rzVrDLFcHq2p\n8TpC+KrfWkjGumKlKdg4ZVYUNLVvIZZK7/XTlI1TZKuqC9Vjz6L54pVAgUvl9+n52UPsWMzq\n4K5thzMZQKw8VQ70Lm1Fq1wp7+rgVC2f3Rf9YEJ3nlR8KNei+CUnAlTFMm0qZWdV6fR/Hsnt\nVETkmN4/hFn3mR9cCn57Gm83I4gVWmJ1fMlzqZkrAVfgMzMbE/Zlj5r0agkSlr1ZXoBudQZP\nqZccxsavb6Tyi6fo4FNXfSn3F4sGP7zYUr+mHmU94iq/1HAzhSpO3OvzR+WUxYzy1VXUa22v\nToWq/jX+pXL60Pq9INcY9dFT6VMBHLWuYwCxXBGp9RGfj0arMvxqDUrfZFJWFdavRi1fB60N\nMyO5ftVqef8rQwLSjCb8OHPieYognTG9W+NlFby+Tft2mvXeVTj62LsaXi+0FotA7q1Qq9GA\nSpRmE3eKSDNu+gW/ze7OZulS2iNMJZmnGKVSfuk6k0PT/I2Ty4OuVOveYk3XYHqvU4g1TUSZ\nthrWubmgwwmZKkFNMKwLOFJJsY1K0hXF5MXJQym4/37SAwT6NFJCJWbJoOkKi+n0ZTmfcwF3\nFyumAvJmimefmwe27rPyHCvl4jNhzqu/Nqf+eWSb6Tk7mZpvPruR3GCeWHU0mSYZmXwq3Dt3\nvVl3Fit9ss95CVix+ZojKF8e2GxqU25abZZLyVMMCwl7VYr3C4vVW6Cm6av3xErmMzF6FEis\naYlAmXvrFLc9Q4gV1NnnTQc+9oW5WL65binknr6rYXbxkPtE6du2RcVmpfna/bJCEkvNvKV0\nTIvlrsm07i1WelU++hyabdc63uWf7Vavn5ldrllT3ZdV+ZosoMtee5FGFjfnuvK+pjxX72gq\nLK3l+Spp6jLdFWrdu4pp4BLUZ1QDHcCd46yZtpH2WV6uX2Jq4eKJqqjl4cRSjdVljj2tNgRE\nOVU012BniEdFvBQbcCKLo0KnPxyVGm7DNGl5UbIr9VlRyMwCS9lRu1f1+Ysrmr6IgRSDW4G5\nXf8qZBredVr5pIkJKuC1qP9EFgrTZyiyCZzGYc3aoOgcVNbcXGvtO31rkFl1snJ38Wp/fBUr\nN8ET+vrinG5znSbWb/fj7b9/W9b8+x5lfH37u66KVVuTHdQVgSeLJ/R55zkde3/2VlChR0Uf\n7+pIVhc7k8qtzO7oXnWC28XKLv4JSaxe07jifi+0/r9ft87w5ddauT5/qC3/uaqKxivFW6aI\nVyuOV9HjuGOp2Wxt5Z+z1jd6zj2havmPVWXrl+kwU56vr6/UrXWWWF98/J5C0Dq53tzLfx/T\no39/Xtzbyiry56tonB+NbotkrealH1zNjt4kr7N6ZVs0XFN9Hh7d9p3s1Rlk+l59x1jVWieK\ndePv+89bjS+L6724D378Mb98V6zy1cYpsvq45I/5bOHaJHn7YK0ci61v4v0i6MzqjAgsRfAl\nXjGJ9+q21yz7qAv4fHMrkvci0tQvK+Zr1m8W2ffZw6kGa3NJVb+c1Wecs78khPpyuVOGAees\nugRd4hXnTl1l1smjwsSFEasSq7ridzG60BmZndes9NfsRKXdSfhOqq09XiQfbJeuwalD1rnz\nWGF7jvUnLrY3x6rFqo7ifK8Tr5fx9UTiAuLrmmFnb/0tL+wpcdXIdrnkYgU51PFEdSNinUQ+\nKtww5fBTbe6PzzVVtJ/veZWO48ywi/8tDPrrV/oZ2WLs4+p8/dRi3atradS58tlWXSrypyMd\nD7f3rRzrJPQ81p9ZP0r+vsUA9/q+cx6r6MzjfqtDMvcl78qorcP4Kj/qvt6oVKLcpvxfHmyI\nkdmiK2xtPjtdOuLTlaf6vTy92BgVnsVdZ951RcVFkmuPu++71elg6isbeo1XziGIWMcuV+hs\n28pFF0elQe5Cut7Be/XNEWRXXKqcxzq71e91rlAtlR+XxTag7qy8NHR9q3WEqq5HLUaI6ybO\n5kceOhotn6TpXPvQeBPylI1eLORpx3TLv3R8Zkvn7WlOswpXTrq3D1Q8RnMRaetMQ5a8z63Y\na+4NIYtruibfn1AphSTpTqcdWiy7C2juI1Zj0n19H8jHVrVwdpzn8hL9XDFXoe+3T0v2o6Yv\nb1WWt7tP7azH37WVzgnylX/6nUz5unMuVB9RNGz1C6qoZ3fjUakC1/yB940rQdf0U+pCm7qX\nWdHUmzOkqlddn5PPle1jRu7UMUsjnvTFgt6nc4NBXymj3tZmXt1HrNb5KH22gQ+cW1KrbCGV\nAC2M7poNNb9eVkK9XUvjTdeYO5vtrfXDQk1ZQJSiQ6e+DEq5lfk0LRWKD75at/oFVbTFkkPT\nOYaLf4pWW/N3etzXOFOhl4Stz7z2J1e8mWpuJh1M9TBMXgVH3ycoz/IRb3xU/0SGEUsnnbqV\nip+Wq+NEu+HOo6X3cqdJS7CPxWV6vlS6WD09VZx0nW7SBEH2cqAOLx1KvpH/lGNd0OIXVbMi\nx3J5H9hqo24LruxfDoaIZlmLZYpW9Nf2zQgh0LUu8e/4nzImR7m604MhR3l5DFkudN7Ndoww\nKuS0vdUTrkjF16xwRLZqiLhgcN1R+vSvUqvR/ZZvL0oQ+BjReRlxLtTTN0H9VDEV0m0OC0aY\nx+L4RYd17dUNxSdZGu29MJ73a+Ta1tlycPJN4/PtufXz/S3MznIF/TgdMKf6vMZRpi+W5ytl\nuq1hwL3EKl7n46WPaX9U6KkNW5+oL8PL4fMvdenLJxSzN0XD+KBu04uhjFWysLwPVR+ojmB7\nnrMY+BnPiBYMIZY6dkUrzWjFg0AVsooB13LI6zM7YTtdbl9sW6Pry4VSg7hWLVqsoN9kTjo8\nfheuNKQY+LnLtBpGrOneFV4tTMa3vmOm9ZHimYtk5iuYrTyWmoccXa4v+sIwE+zUV51S90fH\nhl4I+gTypYbsZACx+HRopdK8WTJy7wjSGtkvsm5h/bMU8r1CfEWrK9K/FFrlJ6CqUwxlki7H\njv7QYl2WKe1nGLG4ifqpRqmAiNWe9F4aVtZ5ji59puJYOS9Bc0rcM1M36B3vUvyRLz1jLpk4\nbUSci3KOPumg5zUh1o4q5EDLGYc1UArfyPF9tsBMEYt1NWJLDD7T8+RUnDiaHvk4pgjKzeBS\nJkkjf8f7y4M1lx0IXoCPXza9Ob5XQ4iV51hhW8jqfh2ba+fYmTSdmkIesILr/nHbO5KMfr+1\n/IgHheEpyGUTdoFnxGWpkB+KIAvUf4zM3cVSR35F8l7EpRgd+mfffP2FoXkNgUKOXqkcmjZW\n49Ud/fSljP1ugSuw78kg+mtamBal/k1EaXd1TpvkHkGru4ul3q70XiUlOi401MpfaJyFo9BR\njQ1oI9Rpm8AlpMVvnngv8tFPPBdXqGQpXaD5bploiAvFhb1LQ+Hq0KiI9eDcW6xyiY5OXcvK\n75VtXrvHv3fqdfcU5VGV8kAv8B31UXpqM8QfsJf0K6S+K+bqnkeFgT6moEKV+oU5pZFTu58/\nsfYYD6fifcWio6inZ1y7V2tqpUKWL1/J7kOSLlSFp3pl8dxtn1rMq8XzKV3OvqNungWPXwHK\nJxZihykfXuDCsqzJ6eORH6u+OyMmXvcWS49+UtPJoKvjk9KGs/PiHEpmT+AeMbTF4t/iChKF\nZGVyXW9okIbWn4CJ31Ssdpvy8xT+8qxcHx5eq+PIrDsjDhXvLpYsIvmWz+NWNdjP7PG+lWjp\n+5B7Kj+8RckPj+dIJeoOxTAVsUgMGXF4+u36UN5TnIoX3TW0qtP1cgH9cvtQtjP+OzNIjuV0\nwkHNV4iUBxp9hiSODYNaQZ9MyU/9ZBcO0omZ5q99RfPix1moR8w6KnaWL33K9tbxnlFWli78\nbB6d+YaYXQpiNV6jKKU7gy2nYNgZr/7OutI0FZ7mL7WncaIitXj6VDAFMdXTpVFhbkzaeApV\nXu1LuW8pLntfHwgnxUKsU6uQkbg6PFt/8ch7uaxJwlgJzyWpp6Z5gVSn41kP53LP06XA1Q45\nx9/NL5c+ZXunymsciGxkuO4YzvWFQ3l1b7HUMZHMJc0ilHboiarkj6dQVHRvvtHfpUBWXuvg\nWY6oGkWYLJuO28iWSI5Fys18LMH1vFLFLA/pZt3BqLDxohwTedgPOzTHoJ1Jp+VodqhQberf\nKAPzaiFO7GPIcqIAZdzxseR+so1ym+LdwhFoNrwKfMutsODOmiKu5d5iqWMiPVQ2Jiy88ipk\nJaNCcYEf/xqvPFvMi3JN09q67hSPXJD5BT0UJPMcdeDp18sWj0FrBmHd8ZkrYmDuL5ZeUrV9\noKtaiqmC4lo+vjxZer+gx4Hl/IMYRY/4XJ+8kM9UyStp6sDJW2G62fnxvCFz7tMYRCyeZWy1\nfSZW9XSaxuIfHXXpJw9DsWLt1e2Lxxz/pCRLFJNySck5r5JZEUm8joSREXPu0xhCrDg9wHOI\nTaWiHT45oj3JFg/UjYlQMaEKsiS/5jnn4s10TvVzZBb9yYPDUqd9fo2Yc5/GGGIpReIqHbGK\n0V5SSXePIYWs6BcnY/Ky8jHfOtXt8cfSvX5JMnmX79R+QQ4FvLEZQSz5rqYqj6/Visl73Wmm\nZk+jf5qQSAvLh9IlRXI6W0rbSQs0xXISUAux1OqHj9azMIRYagorraHmKptu6exbK5ZmuXmO\n9eaGl4l4li8EmZXi7VRnoItZhKybdPJMtn9P3bNtZgyx+J5TrFwpmXdy+aNaLCdixYufQj6j\nEWT+wFWhh153xSyCcqbOuPPRHcSKjCBWWsLrzogmG0qBtEb806F6HSozjTN1Zyb3etks9lBx\nZTep3Uw38ox6YdXufgvGEGsagXGj5m6p2U1q93LuoMieIvkHzJVEEtlkmoEWUr1voKSMN1KH\nPd3v5YJCrIkhxCJX6MxLXMuVeK/Tdq9MjGWIEZHiJzYp0IRAE1OhECvQs07NKrjyodopzg3V\naxBrYhSxlC3ZEy57KZt2p2G/iMjJd6cO2RqxyhXLOBWnxL7m3ITc09as29tvwRhiiUlpXqAe\n9KmPvUjfJ8mOmOZmK1QJUTkqlO0QmbL70L+nMjEqJIYRy2s/BOn52CyZFCizHrZscWM4Ls4t\ntlGsufK+HyOJNd1zlCpmHWiKU6bb8+hTxK6ljen3W45TejXjVYmEfm+BUcRSmRInSroTTB+x\nyb/qltbkfJy60cWtoXXam0FlVHm7q5Yrsf0q4gdiLLFUvFLDfg5ZLp3+Ya3oQ3oyPOSB30I9\nvX5L5fbOte3iglpahbDi6qzvwEBicSzSs0n8FJ30U9NQ8VNX9AkZteh8jXOvVzmU6gVXJFBe\n3X5zhhFLmZWNClWA8umDLtTa+oSesnGmxsWOqpGUt/P0dvHF/XdmJLFyrYIyTI/7eJTG32Me\n15aqehWu6KguFGtNBHxgxhFLIpG6Es8FDlmBjKKQpL4gv58JadZ0VA01Z23NYuAGsZ5+0mss\nsYLO32MGJRk8OxakJ4x9oXrzz8SBfrNn6wfV4FRzz4EyBq7PsZ5+umIwsVRmFa934VGfHiCm\n5aUPXPXW74lVlKDT9TA3gAy1SKtHhRv61wdlJLG0VqpR6aVqcfmyqVWld8XqbeViVGmUuHIe\nC2JdWkUxDoxDQZU9VWTXui/S7qi6bbzc+PtHgRDr2irUaJACEZ9D7K4SetqVtDuqu4iFHOva\nKlS8UvGlDkqeO0DuKVeV3+qoDoh1YEIUo8Jrq1DTWDq+5M2Qvtwj6LnQA/uxP8c6dAoH81hX\nVsGzodMMaOvC4jBFCEe3Z4jVCx5rogpOOncYTCyVv9clyJkcud/WFfYrrZ/03j97VLFkNLE4\nZlViOblyVIlllazgOoWDjCdWChPaFZduYoDyWeTaMN2wBVyncJDhxArySZmsCPrnQvr1bOoR\nTforXKdwlAHFivEq08XJh+NdmmxfP+O+B4h1lPHEKi7dlCeDDAPlm2by108DYh1lRLEkWy+L\nkY835CWfn8EjxzrIgGJxZ1hcFiXyyPUH2f2pYmFUeIxBxVKXXennHTuXFX3CLGkNpj4PMaJY\n6lLS5muVSSZigUMMKRblWP2sKe/7INZ4DCrW/Pd7VNn6+TkWOMigYjWm38tFinkui/M6YD9j\niiWf5lq7Mk4XD8bIYrkNYoHBGFSs2YEheABGFQv5+IMzrlh0LvrcjQEXMaxYbBXUekgGFgvd\n4SMzsliYUH9gIBYwYXixkL8/JiOLFdT3uIMHY2yx0tfNQKzHY2ixMP/+uAwu1vEiwH2AWMCE\nscXCFOnDMrpYOKnzoAwuFiYbHpXhxQKPCcQCJkAsYALEAiYYtbrT2FQBhgYRC5gAsYAJEAuY\nALGACRALmACxgAkQC5gAsYAJEAuYALGACRALmACxgAkQC5gAsYAJEAuYALGACRALmACxgAkQ\nC5gAsZkmJw4AAAV0SURBVIAJEAuYALGACXtb/e3dvArwyOz/dv/V3wIDsb4j+8X6B7FAn72t\n/sut/rAzxPqO7G31z1eIBWY4/HV7tlWARwViARMwjwVMgFjAhEOt/v4DyTtoc6TV3zEqBD2O\ntPqL+21dBXhUMCoEJhxp9Vf3aV0FeFSOtPq/l59/jasAj8qxrhDJO+hgJJY+j/i//4WA/4/9\n/1qxAOgCsYAJEAuYsP8KUvyqCZgBYgEToAQwAWIBEyAWMOGwWMivQAuIBUyAWMAEiAVMgFjA\nBIgFTIAWwIT7ilWdDPp6wnsnf9xe91/P3B5Mr/G5pKDPJ+Gk0nDcs0EmG7QSk1WklrYnPT25\n5UrqcsD9uatY5Ra4cFMq3n7dTbcxWt2c8mFSL9nmk1IkFjr1sbhjc7jinoyK99Gr4NO1E5NJ\nIRpF11ZMz8bvFszKAQMAsYAJEAuYgBwLmIBRITAB81jABDQIMAFiARMgFjABYgETIBYwAWIB\nEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjAB\nYgETIBYwAWIBEyAWMAFiARMgFjABYgETHkes6ltmwAru11x3q3krF20pqnn0mrfyXE3xXNUM\nVfNWnqspnquaoWreynM1xXNVM1TNW3mupniuaoaqeSvP1RTPVc1QNW/luZriuaoZquatPFdT\nPFc1Q9W8ledqiueqZqiat/JcTfFc1QxV81aeqymeq5qhat7KczXFc1UzVM3gqYFYwASIBUyA\nWMAEiAVMgFjABIgFTIBYwASIBUyAWMAEiAVMgFjABIgFTIBYwASIBUyAWMCEBxHr7cW9vH1a\n13LJF2n8pvJt94mqudeXgzyGWD+ng/PDuJaPKxrhg8q33Seq5pJ9avEQYv11Lx/h48X9ta3m\nw73aVnCr4yU1su0+cTVX7FOThxDrzf35uv3PvdtW89u6glsVP1OLm+6TVHPBPrV5CLFe3b9w\nwZvvt/ttW8HX4X4LqcVN90mquWCfOptwn2q3kQ6Sdabw6v78+sqnLav4KHfGZp+kmgv2qQ3E\nEl5jnvvTtpYrxApKrCv2qVX/5TXu4CKxnPsvhM83487jWrGu2adW/VdXuIeLxIp8Gk9rXCtW\nxHqfWvVfXeEeXq4Uy7qaVLz1PuUFXz+R9RBixRHUv4umZK4Ry3qfINYK3qc5nz/OeHDz4m4n\nWKz9TW1svU8cGK/Yp1b9V1e4h4tm3t9urfwZZy7tuGTmnau5Zp9a9V9d4S5+XDJm/nyZqjGO\ni9QrGe9TquaafWrVf3mNe/icrgS4pJof1gNzEst4n3Q15vvUqv/6KsF3AGIBEyAWMAFiARMg\nFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIB\nEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiARMgFjABYgETIBYwAWIBEyAWMAFiHeXFvdx7\nE0YEYh3kj3PXf3PsAwCxDvLLvblf996IAYFYB/nqCF9wEGtwTI7xn3sLb7cfQgI5EOsYP93f\n8PcOv9o2PBDrEJ/TkDD+rgjQQKxD/Df95gP6whqIdYgf02/hfFz/c4DDA7GO8M8R/+69KaMB\nsY7wzmK933tTRgNiHeFHilT/0BeWQKwDfPDvS/50H3fdkvGAWAeQ35c0//XXhwNiHeDlpfUQ\n3IBYwASIBUyAWMAEiAVMgFjABIgFTIBYwASIBUyAWMAEiAVMgFjABIgFTIBYwASIBUyAWMAE\niAVMgFjABIgFTIBYwASIBUyAWMAEiAVMgFjABIgFTIBYwASIBUyAWMAEiAVMgFjABIgFTIBY\nwIT/A7mmTnV1jp15AAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "options(repr.plot.width=5, repr.plot.height=5)\n",
    "plotMA.CD(CD, samplesA = \"Untreated\", samplesB = \"Treated\", col = c(rep(\"red\", 100), rep(\"black\", 900)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are many options for assessing and graphing the results of DE analysis in baySeq.Further, there are many more functions and many more specifications to functions than are used here in order to show a simplified example of one of the tools we use.\n",
    "Obtaining specific, actionable, and publication quality results from analysis requires a deeper understanding of your specific data set and we would love the opportunity to discuss these options with you. \n",
    "\n",
    "While we encourage clients to reach out prior to sequencing so that we can collaborate to design the experiment to answer your specific questions, we look forward to hearing from you at any stage of your RNA-seq project. Please find our contact information available here https://www.cores.emory.edu/eicc/about/index.html"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Session information available below "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 3.6.0 (2019-04-26)\n",
       "Platform: x86_64-w64-mingw32/x64 (64-bit)\n",
       "Running under: Windows 10 x64 (build 18362)\n",
       "\n",
       "Matrix products: default\n",
       "\n",
       "locale:\n",
       "[1] LC_COLLATE=English_United States.1252 \n",
       "[2] LC_CTYPE=English_United States.1252   \n",
       "[3] LC_MONETARY=English_United States.1252\n",
       "[4] LC_NUMERIC=C                          \n",
       "[5] LC_TIME=English_United States.1252    \n",
       "\n",
       "attached base packages:\n",
       "[1] parallel  stats4    stats     graphics  grDevices utils     datasets \n",
       "[8] methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] tseries_0.10-47      future_1.14.0        baySeq_2.18.0       \n",
       " [4] abind_1.4-5          GenomicRanges_1.36.0 GenomeInfoDb_1.20.0 \n",
       " [7] IRanges_2.18.1       S4Vectors_0.22.0     BiocGenerics_0.30.0 \n",
       "[10] edgeR_3.26.6         limma_3.40.5         BiocManager_1.30.4  \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] Rcpp_1.0.2             pillar_1.4.2           compiler_3.6.0        \n",
       " [4] XVector_0.24.0         xts_0.11-2             base64enc_0.1-3       \n",
       " [7] bitops_1.0-6           tools_3.6.0            zlibbioc_1.30.0       \n",
       "[10] zeallot_0.1.0          digest_0.6.20          uuid_0.1-2            \n",
       "[13] tibble_2.1.3           jsonlite_1.6           evaluate_0.14         \n",
       "[16] lattice_0.20-38        pkgconfig_2.0.2        rlang_0.4.0           \n",
       "[19] IRdisplay_0.7.0        curl_4.0               IRkernel_1.0.2        \n",
       "[22] GenomeInfoDbData_1.2.1 repr_1.0.1             vctrs_0.2.0           \n",
       "[25] globals_0.12.4         locfit_1.5-9.1         grid_3.6.0            \n",
       "[28] listenv_0.7.0          pbdZMQ_0.3-3           TTR_0.23-4            \n",
       "[31] backports_1.1.4        codetools_0.2-16       htmltools_0.3.6       \n",
       "[34] quantmod_0.4-15        quadprog_1.5-7         RCurl_1.95-4.12       \n",
       "[37] crayon_1.3.4           zoo_1.8-6             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
