# Methods
A selection of the analytical processes and tools we use to provide rigorous, reproducible, and actionable results to our clients. If you are interested in a project with EICC please contact us via email at eicc@emory.edu or call us at (404) 778-5123

# RNA-Seq Differential Expression Analysis

# baySeq
A demonstraton of the R/Bioconductor package baySeq to perform differential expression analysis from RNA-Seq count data. baySeq uses empirical Bayes methodology to estimate prior and posterior likelihoods of differential expression between comparison groups. View the notebook in the link above or click here https://nbviewer.jupyter.org/github/EmoryIntegratedComputationalCore/Methods/blob/master/RNA-SeqDEAnalysiswithbaySeq.ipynb?flush_cache%3Dtrue

# DESeq2
A demonstraton of the R/Bioconductor package DEseq2 to perform differential expression analysis from RNA-Seq count data. DESeq2 primarily uses the Wald test to determine differential expression between comparison groups.

# edgeR
A demonstraton of the R/Bioconductor package edgeR to perform differential expression analysis from RNA-Seq count data. edgeR primarily uses the quasi-likelihood F test to determine differential expression between comparison groups. View the notebook in the link above or click here https://nbviewer.jupyter.org/github/EmoryIntegratedComputationalCore/Methods/blob/master/RNA-SeqDEanalysisedgeR.ipynb

# 16s Microbiome Analysis (coming soon)

# Descriptive and Statistical Analysis of 16s microbiome data with phyloseq and LDM (coming soon)


# Miscellaneous

# False Discovery Rate Estimation with qvalue
A demonstraton of the R/Bioconductor package qvalue to estimate global and local false discovery rates from p-values and test statistics calculated from  differential expression analysis.

# Power & Sample Size Assesment for RNA-Seq with PROPER
A demonstration of the R/Bioconductor package PROPER which uses simulation of experimental design constraints to provide guidance on how many people each comparison group needs for sufficient statistical power. View the notebook in the link above or click here https://nbviewer.jupyter.org/github/EmoryIntegratedComputationalCore/Methods/blob/master/Power%26SampleSizewithPROPER.ipynb?flush_cache%3Dtrue
