---
title: 'RNA-seq Differential Expression (DE) Analysis Using edgeR'
author: "Jessica Randall"
output:
  bookdown::pdf_document2:
    fig_width: 5
    latex_engine: xelatex
    toc : false
---

```{r logo, message = FALSE, echo=FALSE}
knitr::include_graphics("eicc_logo.png")
```

Briefly, edgeR uses a negative binomial distribution to determine the likelihood of differential expression between comparisons of two or more groups. This distribution mathematically accounts for the fact that we are assessing gene counts and we are assuming that most genes we are comparing between the groups will not be differentially expressed. The package is flexible in that it provides the user options to use empirical Bayes methods or frequentist methods which rely on the exact test, linear models, and quasi-likelihood tests.

Linked is the [original 2010 paper](https://www.ncbi.nlm.nih.gov/pubmed/19910308) from Robinson, McCarthy, and Smyth introducing the concepts implemented in edgeR. 

There are two additional papers exploring improved functionality. One from McCathy, Chen, and Smyth in 2012 on accouting for [biological variation](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3378882/) and one from Chen, Lun, and Smyth in 2016 on establishing an [example workflow](https://www.ncbi.nlm.nih.gov/pubmed/27508061) for package functions 

We will be using the [pasilla package](https://www.bioconductor.org/packages/release/data/experiment/html/pasilla.html) for our example data.

edgeR has wide range of applications to genomic analyses. Our simplified example describes the quasi-likeliehood F-test which is the method in edgeR which provides the strictest control of potential false positive genes. We strongly encourage you to reach out to EICC with questions regarding options available to you with edgeR. Check out some of the graphs from previous projects [here](https://github.com/EmoryIntegratedComputationalCore/Methods/blob/master/DataVisualizationMenu.pdf).

### Definition of terms {-}

#### Quasi-likeliehood F Test: {-}

Unadjusted p values vs adjusted p values/(FDR): In DE analysis, a single p-value tells you how likely it is that a single gene is differentially expressed between at least two groups (ex: a control and a treatment group) due to some actual difference between the groups as opposed to random chance. False Discovery Rate (FDR) tells you how likely it is that all genes identified as DE are false positives. A FDR of 5% means that among all genes called DE, an average of 5% of those are truly not DE.DE genes are only considered significantly so if they meet the adjusted p value, not only the unadjusted p value. 

Compared to DESeq2 it has been our experience that edgeR is more stringent in its calling of genes as significantly DE. If your study is exploratory and are not sure which or how many DE genes you are expecting to find in your experiment, edgeR may not be the best package for your analysis.

#### Unadjusted p values vs adjusted p-values/(FDR): {-} 

In DE analysis, a single p-value tells you how likely it is that a single gene is differential expressed between at least two groups (ex: a control and a treatment group) due to some actual difference between the groups as opposed to random chance. False Discovery Rate (FDR) tells you how likely it is that all genes identified as DE are false positives. A FDR of 5% means that among all genes called DE, an average of 5% of those are truly not DE. DE genes are only considered significantly so if they meet the adjusted p value, not only the unadjusted p-value. FDRs for each individual gene are called q-values or local FDRs. 

### Loading data {-}

Our very first step is to load the libraries we'll need to assess the functions required for analysis and graphing. Please see [Bioconductor](http://bioconductor.org/) for information about initial installation and use of Bioconductor and its packages. We also set the minimal theme in gglot2 for all graphs to have the same aesthetic features by default.

The pasilla experiment studied RNAi knockdown of Pasilla, the Drosophila melanogaster ortholog of mammalian NOVA1 and NOVA2, on the transcriptome. Data are provided by NCBI Gene Expression Omnibus under accession numbers GSM461176 to GSM461181.

Here we will demonstrate importing the count matrix and sample data from the pasilla package since we're using it as an example. Typically we will use the here package to specify the path for the counts and sample data files in a list of files to import and export from the task. 

We're also going to specify that we'd like the row names of our sample data to come from the first column, called "file" since this is where we've stored which sample is which and finally we remove extra columns from our sample data which we won't be using in our analysis.

Please reach out to EICC if you would like to compare 3 or more groups as this is a simplified example. It may also be the case you will need more than 6 samples per experimental group or that you may need to remove genes with average counts greater than 5, 10, 15, or even 20 for sufficient statistical power. Please see our PROPER walk-through for an example of our of power and sample size analysis.

```{r load libs and data, message = FALSE, echo = FALSE}

require ("pacman")
p_load("readr", "dplyr", "knitr", "tidyverse", "edgeR", 
       "ggplot2", "EnhancedVolcano", 
      "org.Dm.eg.db", "AnnotationDbi")

theme_set(theme_minimal())

countdata <- as.matrix(read.csv(system.file("extdata", 
                                            "pasilla_gene_counts.tsv",
                                            package="pasilla",mustWork=TRUE), sep = "\t", row.names = "gene_id"))

sampledata <- as.data.frame(read.csv(system.file("extdata",
                                                 "pasilla_sample_annotation.csv",
                                                 package="pasilla", mustWork=TRUE), row.names = 1))

sampledata <- sampledata[,c("condition","type")]

```

### Preparing for Analysis {-}

In order to preform a pairwise comparison we need to specify some information about our data. In edgeR we must create a special object called a DGElist object, here abbreviated as y. 

This object takes in the countdata object and the sample data comparison variable of interest as a vector to create the DGEList object and prepare for analysis. 

Since we want to know if edgeR filters out any additional genes with low counts, we create another object called y2 to perform additonal filtering. We save the unfiltered DGElist object as y so we can check how many genes we are starting with and how many are filtered for low counts later on. 

Since edgeR looks at data alphabetically, we also need to make sure we specify the untreated group is our reference group with the ref="untreated" statement.

``` {r create DGE List object, message = FALSE, echo = FALSE}

y <- DGEList(counts = countdata, 
             group = sampledata$condition)

y2<-y

y2$samples$group <- relevel(y2$samples$group, 
                           ref="untreated")

```

### Annotating gene labels {-}

Typically we receieve and provide results using genes labelled with gene symbols. However, occasionally someone will prefer Entrez or Ensemble IDs. This is an example of how we would annotate genes providede with Ensemble IDs with their corresponding gene symbols.

Annotation is done using the Annotation Dbi package and the annotation library specific to your organism, in this case, we use org.Dm.eg.db for Drosophila melongaster. This allows us to annotate genes with different labels regardless of the pacakge used for differential expression analysis. 

Please note that all IDs which no longer have official gene symbols will be removed from the analysis at this step. In some cases, IDs may map to many values in another labelling scheme and in this case we have selected to only keep the first label. 

```{r annotate, message = FALSE, echo = FALSE}

y2$genes$Symbol <- mapIds(org.Dm.eg.db, 
                         rownames(y2),
                         keytype="ENSEMBL", 
                         column="SYMBOL", 
                         multiVals = "first")

y2 <- y2[!is.na(y2$genes$Symbol), ]
                         
```

If we had any additional data to add about the samples that we wanted to include in our analysis we would add it next but since this is a simplified example, we are only comparing treated and control samples without taking into account any additional information about them.

At this point we generate our first exploratory visualization, the principal components analysis plot. This will show us how your data cluster or how similar each sample is to others of the same group.

By default, edgeR produces a general multi-dimensional scaling plot that tells us how closely related the samples are to each other by log fold changes. We can specify that we would like to view the special case of a PCA plot which tells us how much group membership (i.e. being untreated or treated) contributes to the differences between the samples and we hope this will be high. 

```{r exploratory pca, fig.height=7, fig.width=7}
colors <- rep(c("red", "blue"), 2)

plotMDS.DGEList(y, 
                col=colors[sampledata$condition], 
                gene.selection="common",
                main="Variability between Samples by Condition")

```

This is a great looking PCA plot because it shows that it is our samples are clustered by so we can say that the experimental condition is largely responsible for the variation between the samples rather than noise generated by sequencing, analysis, or some other unaccounted for variable.

I might also say that the within-group variability between samples in the treated group is probably contributing some noise to our ability to detect differences between the treated and untreated groups. The y-axis tells us how much variability between these samples is due to other factors in our model or if we have none, sources of variability we may not have accounted for like sex or ethnicity which are often leading contributors of variability between samples and should be accounted for in experimental design if you wish to control for their effects. 

### Prefiltering {-}

Typically we want to ignore genes that have counts of zero across all samples since these are adding statistical noise. We may also want to be more stringent and remove genes with rows that sum to 10, 20 or even 30 or less since these could also be contributing statistical noise that would prevent us from seeing the differences between the cases and controls, especially if it is a small one. Think of it like removing the static from an analog TV. Once it's gone the picture is more clear.

For this example, we will use the filterbyExpr function to filter on counts per million (cpm) and let edgeR choose the filtering cutoff for us based on the data. We can even check how many genes were filtered at this step. We see that 6680 genes were filtered out and 7919 were kept in for analysis.

If you choose to use EICC and edgeR specifically to do your differential expression analysis, we do have the option to customize this part of the analysis based on your data and are able to filter more or fewer lowly expressed genes as needed. 

```{r, filter, message = FALSE, echo = FALSE}

keep <- filterByExpr(y2)

table(keep)

```

### Recalculate library size {-}

Since we've removed genes with low counts we need to recalculate the library size before we normalize it for analysis. DESeq2 does this sort of behind the scenes from what a typical user would see but edgeR requires the user to do this by hand.

```{r recalc, message = FALSE, echo = FALSE}

y2 <- y2[keep, ,keep.lib.sizes=FALSE]

dim(y2)

```

### Normalizing data {-}

There are a number of reasons we may want to normalizing data in edgeR. We want to account for gene length, GC content, sequencing depth, or sample specific effects. edgeR minimizes the effects of most of these concerns behind the scenes but we do have to do a little work ourselves to account for RNA composition. If you choose to do an RNA-seq project with EICC we would choose normalization strategies customized to your data.

The calcNormFactors function here and the trimmed mean of M-values (TMM) method specifically is used in most, if not all differential expression analysis because we assume that the majority of genes are not going to be DE. This function uses the original library size from the raw counts to create an effective library size which is scaled by the log-fold changes between the samples for most genes. This scaling prevents or at least minimizes the effects of very extreme log-fold changes skewing results. 

As you can see from the graph below, the library sizes of the samples vary from one to the next. In order to compare them, we need them to be approximately the same, similarly to how log transforming data scales it to help it look more like a normal/bell/Gaussian curve, normalization with TMM puts these varying library sizes on the same scale so we can compare the samples. 

```{r normal, message = FALSE, echo = FALSE}

barplot(y2$samples$lib.size*1e-6, 
        names=1:7, 
        ylab="Library size (millions)", 
        main="Library Size")

y2 <- calcNormFactors(y2, method="TMM")

barplot(y2$samples$lib.size*1e-6, 
        names=1:7, 
        ylab="Library size (millions)", 
        main="Library Size")

```

### Estimating biological variability between samples {-}

In DE analysis we are testing the idea that gene expression varies between these groups, control and treatment, and we expect that most of the variation between the samples comes from them being either a control sample or a treatment sample. However, there will always be a certain amount of variability we cannot anticipate but that we can measure. This inherent biological varaibility, also called the biological coefficient of variation or BCV, also varies by the type of sample. In well controlled experiments technical replicates would have smaller BCVs than unrelated, unique individuals.

First, we estimate the dispersion which is the amount of variance in our experiment. This is critical to the analysis because we want genes that appear to be consistently high or low across samples to count more in analysis than genes that vary (which signals possible outlier genes) count less. We have specified that robust=TRUE that we may account for any outlier genes. If there are none, this does not affect estimation. 

We then take the square root of the dispersion to give us the BCV. This check of the BCV serves as a measure of quality control for the analysis. Since our samples are from genetically similar samples, we would expect our BCV to be between 0.05-0.2.

```{r, bcv, message = FALSE, echo = FALSE}

y2 <- estimateDisp(y2,
                   design, 
                   robust=TRUE)

bcv <-sqrt(y2$common.dispersion)

bcv
```

Our biological variability is estimated to be 0.14 which makes sense for genetically similar individuals.

### Testing {-}

At this point we have prepared all of the data to be input into our quasi- likelihood F (QLF) test. Recall that the hypothesis we are testing is that there is no difference in gene expression between the treatment and control samples. We chose to use the QLF test because of all of the tests edgeR offers, it is the most up to date as of writing this and gives us the strongest control of false postive results. In this step we specify our comparison of interests, treated to untreated, we fit the QLF model, test it, and summarize our results.

Since in our example we have an idea of what we're expecting to find in terms of the number of DE genes, we want to be sure we have as few false positive DE genes as possible so that we know whichever we do see, we can confidently follow up on with a lab test. 

The decision to use edgeR will vary by your experimental goals. If you use edgeR, the QLF test is always the recommended test because even if you are doing a simple pairwise comparsion now, you can add in additional variables like sex, ethnicity, or batch with relative ease as compared to the other tests available. Please contact EICC if you would like to discuss your experimental goals so we can assist you in choosing the appropriate analysis tool.

``` {r test, message = FALSE}

comparison <- makeContrasts(treated-untreated, levels=design)

qlfit <-glmQLFit(y2, 
                 design, 
                 robust=TRUE)

qlf <- glmQLFTest(qlfit, 
                  contrast=comparison)

summary(decideTests(qlf, adjust.method="fdr", p.value=0.05)) 

```

### Results {-}

In this example we see that while controlling the adjustedp/FDR threshold at <0.05, we have 695 (9.6%) DE genes. Of these, 358 (4.9%) are up regulated and 337 (4.7%) are down regulated while 7224 (91%) are not significantly differentially expressed.

In our walkthorugh of DESEq2 we use the same example dataset and found that while controlling the adjustedp/FDR threshold at <0.1, we had 1,054 (10.6%) DE genes. Of these, 518 (5.2%) were up regulated and 536 (5.4%) were down regulated. There was 1 outlier gene (0.01%) detected and 1539 genes (16%) with low counts were removed for having counts <6 across all samples. Different analytical tools will often give you slightly different results and edgeR may be better for projects which have specified genes of interest in mind rather than exploratory projects since edgeR is much more strict with potential false positive results while DESeq2 may allow more false positives into your significant results but it provides you additional tools for evaluating their veracity before following up with a lab test.

FDR thresholds can also range from 0.05-0.2 and it is much better to have the option to lower your threshold during analysis rather than have to increase it because your study was not sufficiently powered to find anything but the most extremely highly or lowly differentially expressed genes. Please reach out to us at EICC if you would like assistance in planning your experimental design for your RNAseq project and in setting appropriate FDR thresholds.

### Visualizing {-}

We now perform additional data visualizations. Typically we provide a PCA plot, heat maps, and volcano plots. We would be happy to work with you to customize these for publication. Please see our Data Visualization [menu](https://github.com/EmoryIntegratedComputationalCore/Methods/blob/master/DataVisualizationMenu.pdf) for more options and examples from previous projects.

Before we generate our heatmap in edgeR, we need to transform the counts in the DGEList object y) into counts per million (CPM) and then log transform them. In this example we have decided to sort the DE genes by smallest adjusted p value and only examine the top 20 DE genes. The number of DE genes you may want to visualize is customizable based on your project with EICC.

```{r visualize, echo= FALSE, fig.height=3, fig.width=3, message=FALSE}

logCPM <- cpm(y2, prior.count=2, log=TRUE)

rownames(logCPM) <- rownames(countdata)

colnames(logCPM) <- paste(y2$samples$group, sep="-")

o <- order(qlf$table$PValue)

logCPM <- logCPM[o[1:20],]

coolmap(logCPM, margins=c(8,8))

```

Please note that these are sorted for convenience but the gene at the top of the list is no more significant than the gene at the bottom of the list. As is the case with nominal p-values, a smaller adjusted p-value does not make a gene more statistically significant than one with a larger adjusted p-value. If the genes are below the threshold, they are all equally statistically significantly differential expressed. These are sorted for convenience but the gene at the top of the list is no more significant than the gene at the bottom of the list. 

We also typically provide clients with an initial volcano plot created with the EnhancedVolcano R library. Similar to the PCA plot and heat map, this is a highly customizable graph and we would like to work with you to design graphs which best tell the story of your results. 

A volcano plot is technically a scatter plot where the x-axis has the log2 transformed fold changes between the compared samples and the y axis has the local adjusted p-values for each gene, also called the q-value. Here we have also labelled the genes with FDR < 0.1 as that is where we set our threshold when we generated our results. The points in red are those which meet the threshold for statistical significance with a q value  less than or equal to 0.1 and a log2 fold change of 1.0 or greater. Points in green are those with only log2 fold changes >1.0 and those in blue have claques < 0.1. The points in grey are non statistically significant by any measure. All of these parameters can be adjusted based on your cutoffs and thresholds. 

```{r volcano, echo= FALSE, fig.height=5, fig.width=5, message=FALSE}

EnhancedVolcano(y2,
                lab = y2$genes$SYMBOL,
                x = [log2 fold changes],
                y = [adjusted p-vals],
                xlim = c(-6, 6),
                title= NULL,
                subtitle= "Log(2) Fold Change vs -log(10) q values",
                FCcutoff = 1.0,
                pLabellingCutoff = 0.05,
                pCutoff = 0.05,
                legendPosition = "bottom",
                legend=c("NS", "Log2 fold-change", "adj P-value",
                         "adj P-value & Log2 fold-change"))

```

There are many more functions and many more specifications to functions than are used here in order to show a simplified example of one of the tools we use for differential expression analysis. Obtaining specific, actionable, and publication quality results from analysis requires a deeper understanding of your specific data set and we would love the opportunity to discuss these options with you.

While we encourage clients to reach out prior to sequencing so that we can collaborate to design the experiment to answer your specific questions, we look forward to hearing from you at any stage of your RNA-seq project. Please find our contact information available on our [website](https://www.cores.emory.edu/eicc/about/index.html) and check out some of the graphs we've made for previous clients [here](https://github.com/EmoryIntegratedComputationalCore/Methods/blob/master/DataVisualizationMenu.pdf).

### Session information and References {-}

```{r sessioninfo, message = FALSE, echo = FALSE}
date()
sessionInfo()
citation("bookdown")
citation("readr")
citation("knitr")
citation("ggplot2")
citation("edgeR")
citation("EnhancedVolcano")
```
<!---- done ---->
