---
title: "RNA-seq Differential Expression (DE) Analysis Using DESeq2"
author: "Jessica Randall"
date: "1/30/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The original paper from Anders & Huber that introduces the concepts implemented in DESeq2 is available here https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8Love, M.I., Huber, W., Anders, S. Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2 Genome Biology 15(12):550 (2014)

We will be using the pasilla package for our example data which is cited here. Wolfgang Huber and Alejandro Reyes (2019). pasilla: Data package with per-exon and per-gene read counts of RNA-seq samples of Pasilla knock-down by Brooks et al., Genome Research 2011.. R package version 1.12.0.

Data for the package is part of this paper: "Conservation of an RNA regulatory map between Drosophila and mammals" by Brooks AN, Yang L, Duff MO, Hansen KD, Park JW, Dudoit S, Brenner SE, Graveley BR, Genome Res. 2011 Feb;21(2):193-202, Epub 2010 Oct 4, PMID: 20921232.

Since we will be using the apeglm shrinkage estimator to create our MA plot, I have also cited those authors
Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for sequence count data: removing the noise and preserving large differences. Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895

Briefly, DESeq2 uses a Wald test to determine differential gene expression between at least two groups. DESeq2 assumes a negative binomial model which mathematically accounts for the fact that we are assessing gene counts and we are assuming that most genes we are comparing between the groups will not be differentially expressed.

Please note that DESeq2 has a number of capabilities that we will not be covering. You can import data for use in DESeq2 in several different ways, there are options for single cell projects, for incorporating Bayesian statistics, for time-series experiments, for outliers, for obtaining all of the results that DESeq2 functions produce, and even more options for graphing (including an R shiny app we plan to cover in a future walkthrough) so we strongly encourage you to reach out to EICC with questions regarding options available to you with DESeq2.

Definition of terms

######Wald test: 
DESeq2 offers two options with how it determines differential expression, the Wald test and the Likelihood Ratio Test. Users of edgeR may be familiar with the latter. In DESeq2, the default test for pairwise comparison analysis is called the Wald Test and it is looking at your control and your experimental samples to see if the difference between them is equal to zero or not. Compared to edgeR it has been our experience that DESeq2 is more liberal in its calling of genes as significantly DE. If you are not sure which or how many DE genes you are expecting to find in your experiment, DESeq2 would likely the best analysis tool for your data.

######Shrinkage estimator: 
Shrinkage estimators help us more accurately measure exactly how much the log transformation is condensing the data as it tends to do, especially when there are very small and very large values in one dataset together. Choosing the best estimator to assess this depends on your experimental aims and what you would like to do with your data but picking one is essential for accurately visualizing data analyzed in DESeq2.

######Unadjusted p values vs adjusted p values/(FDR): 
In DE analysis, a single p-value tells you how likely it is that a single gene is differentially expressed between at least two groups (ex: a control and a treatment group) due to some actual difference between the groups as opposed to random chance. False Discovery Rate (FDR) tells you how likely it is that all genes identified as DE are false positives. A FDR of 5% means that among all genes called DE, an average of 5% of those are truly not DE. DE genes are only considered significantly so if they meet the adjusted p value, not only the unadjusted p value.
Our very first step is to load the DESeq2 package and the pasilla package from Bioconductor. Please see http://bioconductor.org/ for information about initial installation and use of Bioconductor and its packages.

```{r load libs and data, cars}
require ("pacman")
p_load("readr", "dplyr")
```
